README.md

```diff
@@ -1 +1,277 @@
+
+There are currently two main datastore implementations:
+ 1. [`datastore-fs`](https://github.com/ipfs/js-datastore-fs) that is backed by file system and is used mainly in the NodeJS environment
+ 2. [`datastore-level`](https://github.com/ipfs/js-datastore-level) that is backed by LevelDB and is used mainly in the browser environment
+ 
+ Both implementations share the same API and hence are interchangeable. 
+
+ When the migration is run in a browser environment, `datastore-fs` is automatically replaced with `datastore-level` even 
+ when it is directly imported (`require('datastore-fs')` will return `datastore-level` in a browser). 
+ So with simple migrations you shouldn't worry about the difference between `datastore-fs` and `datastore-level` 
+ and by default use the `datastore-fs` package (as the replace mechanism does not work vice versa).
+
+### Guidelines
+
+The recommended way to write a new migration is to first bootstrap a dummy migration using the CLI:
+
+```sh
+> npm run new-migration
+```
+
+A new folder is created with the bootstrapped migration. You can then simply fill in the required fields and 
+write the rest of the migration! 
+
+### Integration with js-ipfs
+
+When a new migration is created, the repo version in [`js-ipfs-repo`](https://github.com/ipfs/js-ipfs-repo) should be updated with the new version,
+together with updated version of this package. Then the updated version should be propagated to `js-ipfs`.
+
+### Tests
+
+If a migration affects any of the following functionality, it must provide tests for the following functions
+ to work under the version of the repo that it migrates to:
+
+* `/src/repo/version.js`:`getVersion()` - retrieving repository's version
+* `/src/repo/lock.js`:`lock()` - locking repository that uses file system
+* `/src/repo/lock-memory.js`:`lock()` - locking repository that uses memory
+
+Every migration must have test coverage. Tests for migrations should be placed in the `/test/migrations/` folder. Most probably
+you will have to plug the tests into `browser.js`/`node.js` if they require specific bootstrapping on each platform.
+
+### Empty migrations
+
+For interop with go-ipfs it might be necessary just to bump a version of a repo without any actual 
+modification as there might not be any changes needed in the JS implementation. For that purpose you can create an "empty migration".
+
+The easiest way to do so is with the CLI:
+
+```sh
+> npm run new-migration -- --empty
+```
+
+This will create an empty migration with the next version.
+
+### Migrations matrix
+
+| IPFS repo version  | JS IPFS version  |
+| -----------------: |:----------------:|
+|                  7 | v0.0.0 - latest  |
+
+## Developer
+
+### Module versioning notes 
+
+In order to have good overview of what version of package contains what kind of migrations, this package follows this versioning schema: `0.<versionOfLastMigration>.<patches>`.
```

Comment:
Tying the npm version to the repo version like this means we lose the ability to communicate breaking/feature change information for _this_ module to dependants.

For example, how would we communicate a change like the async/await refactor in this module? There's no repo version increase and we can't release a breaking change like that in a patch release.

Reply:
Well, you are right about this, but I don't believe there will be much of feature development/breaking changes in the future of this package. I believe that 80% of the changes will be adding new migrations, then 15% will be fixing bugs and 5% others...

In your example, we could communicate the breaking change using the major version. So for example from `0.8.0` we could bump to `1.8.0`... 
I kind of like the idea that you know what migrations the package contains just from the version number.

Reply:
> In your example, we could communicate the breaking change using the major version. So for example from `0.8.0` we could bump to `1.8.0`...

We could...IMHO that would be a really strange versioning strategy though!

> I kind of like the idea that you know what migrations the package contains just from the version number.

I kind of do too, but I'm not convinced it's worth it!

Reply:
> We could...IMHO that would be a really strange versioning strategy though!

True, but following `0.<versionOfLastMigration>.<patches>` is already "strange" anyway ;-) 

What about having two parts in the version? E.g. `<semver>-<last version of migration>` so it could look like `0.2.0-7`? When new migration is added it would be bumped to `0.3.0-8`. This would allow us to have full-fledged semver capabilities and still have the information about migrations as part of the version?

What I am not sure is if this schema would not break some tooling? (aegir?)

Reply:
Is there another reason other than convenience that would warrant this versioning strategy? 

If you don't already know the module the versioning strategy isn't going to convey any additional info. Secondly, it's really easy for new developers to miss this important piece of info since it's a small section at the bottom of the README. Thirdly, if you do know the module and are upgrading to get the latest migrations you'll probably just need the latest version of this module anyway right? It's quite unlikely you'll ever need an older version of the module since the newer versions contain all the past migrations as well.

Reply:
I would not call it convenience, but being careful. Having a different version of this package installed may lead to irreversible actions that can happen without your knowledge (eq. when you have `autoMigrations` on - which is currently the default). Yes, it won't most probably lose the data per se, but it might migrate data that are not to be migrated at that moment (for example you have test repo and "production" repo and you want to do some tests, but you accidentally migrate wrong one etc...).

Of course, having this schema won't prevent all mix ups, but it can improve the situation. From version update `0.4.0 --> 0.5.0` you can't be sure if there was new migration added (yes, it will be most of the cases), or maybe some feature change? You will have to go study release notes, which should be always the case, but I am pretty sure a lot of devs don't do that. 

In this light, I think having the `-<latestVersionOfMigration>` suffix is even better than the original schema as it stands out from Semver and should tingle dev's senses about what it most probably means.

I am not sure if my thoughts about this are not getting too tangled and strayed from reality, but I still feel like there are more pros over cons with this schema. Actually are there any cons? Except for the possible tooling side?

Reply:
If you feel strongly about this I'll not stand in your way any further.

I'm undecided if I agree that a pre-release suffix is a better strategy than using the regular major/minor/patch versions. My issue with this it is that according to semver:

> A pre-release version indicates that the version is unstable and might not satisfy the intended compatibility requirements as denoted by its associated normal version

So this would convey to devs that it is not a stable version that should not be used without accepting some instability consequences.

I also do not know if AEgir will support you on this 😛

Reply:
Hmmm. Agree that this is a bit unfortunate. I checked finally SemVer spec and it directly supports ["build metadata"](https://github.com/ipfs/js-ipfs-repo-migrations/pull/4), which I feel could be exactly used for this. It stands aside in determining version precedence, which is a good thing.

We could have then schema `<semver>+<latestMigrationVersion>` (i.e `0.1.0+7`) to meet both points... Only thing I am not sure is about the format of the metadata. Should it be just integer of the version or somethings like `+migration.<int>`? It would be more self-explanatory, but longer...